### ISSUE TITLE
Test inclusion in package

### ISSUE THREAD

**Issue by antnieszka** [DataMedSci/beprof#111]
Should we include tests in our release package?
E.g. numpy does it
https://github.com/numpy/numpy/blob/master/numpy/setup.py#L23
https://github.com/numpy/numpy/blob/master/numpy/distutils/setup.py
https://github.com/numpy/numpy/blob/master/numpy/distutils/misc_util.py#L1019

Some info from official site: 
data_files in setup.py: https://docs.python.org/2/distutils/setupscript.html#installing-additional-files

I looked into installed version of numpy and tests are included:
```
(beprof3){16-12-05 22:43} ~/.virtualenvs/beprof3/lib/python3.4/site-packages/numpy-1.11.1-py3.4-linux-x86_64.egg/numpy ant6% ➜ ll
total 304K
[...]
-rw-r--r-- 1 ant6 users  920 07-02 18:56 setup.py
drwxr-xr-x 4 ant6 users  185 07-02 18:56 testing
drwxr-xr-x 3 ant6 users  124 07-02 18:56 tests
-rw-r--r-- 1 ant6 users  294 07-02 18:56 version.py
```

Some of them hidden in sub directories like `core/tests/...`, `fft/tests/...`. They are also in system-installed version of numpy (the one above is pip-installed in virtenv)

**Comment by grzanka** [comment]
data_files in setup.py is not meant to be used for storing any python files.

Check packages:
 - https://github.com/DataMedSci/beprof/blob/master/setup.py#L117
 - https://github.com/DataMedSci/pymchelper/blob/master/setup.py#L118

and understand what is in MANIFEST.in file

**Comment by antnieszka** [comment]
Ok, in MANIFEST all *.py files are included and /tests directory is not excluded, so if we added 
```
setuptools.setup(
...
packages=['beprof', 'tests']  # <-- added tests here
```
There is similar example in python docs: https://pythonhosted.org/an_example_pypi_project/setuptools.html#setting-up-setup-py

Or use ` packages=setuptools.find_packages(where='.'),`. @grzanka could You explain why there are two different approaches? Also I see another solution - move tests directory to package beprof/tests/ and it should be included (also change line in MANIFEST then).

**Comment by antnieszka** [comment]
Probably we could integrate tests with pytest-runner:
http://doc.pytest.org/en/latest/goodpractices.html#integrating-with-setuptools-python-setup-py-test-pytest-runner

**Comment by grzanka** [comment]
I'de recommend keeping test package outside beprof (as it is mostly done in many projects).
find_packages allows you to avoid redundancy in putting your package name everywhere, its also very popular approach: http://www.programcreek.com/python/example/10688/setuptools.find_packages

Speaking of pytest-runner - sure, but its a separate issue.

**Comment by antnieszka** [comment]
Ok, I'll add 'tests' package to setup.py and create a separate issue-feature for pytest-runner. After all - it's another dependency...

**Comment by antnieszka** [comment]
Well, I changed the following in setup.py:
```
    packages=['beprof', 'tests'],
    test_suite='tests',
```
And executing `python setup.py test` runs test, but... as I package it with `python setup.py build bdist` it lands in wrong place, or I think so:
```
{16-12-06 15:08} ~/Downloads/site-packages ant6% ➜ ll -R
.:
total 0
drwxr-xr-x 2 ant6 users 195 12-06 15:05 beprof
drwxr-xr-x 2 ant6 users 110 12-06 15:05 beprof-0.post0.dev1821481033107-py2.7.egg-info
drwxr-xr-x 2 ant6 users 135 12-06 15:05 tests

./beprof:
total 64K
-rw-r--r-- 1 ant6 users  11K 12-06 14:49 curve.py
-rw-r--r-- 1 ant6 users  13K 12-06 15:05 curve.pyc
-rw-r--r-- 1 ant6 users 1,9K 12-06 14:49 functions.py
-rw-r--r-- 1 ant6 users 2,4K 12-06 15:05 functions.pyc
-rw-r--r-- 1 ant6 users    0 12-05 23:16 __init__.py
-rw-r--r-- 1 ant6 users  164 12-06 15:05 __init__.pyc
-rw-r--r-- 1 ant6 users 5,8K 12-06 14:51 profile.py
-rw-r--r-- 1 ant6 users 6,3K 12-06 15:05 profile.pyc
-rw-r--r-- 1 ant6 users  478 12-06 15:05 _version.py
-rw-r--r-- 1 ant6 users  578 12-06 15:05 _version.pyc

./beprof-0.post0.dev1821481033107-py2.7.egg-info:
total 24K
-rw-r--r-- 1 ant6 users    1 12-06 15:05 dependency_links.txt
-rw-r--r-- 1 ant6 users 6,6K 12-06 15:05 PKG-INFO
-rw-r--r-- 1 ant6 users   14 12-06 15:05 requires.txt
-rw-r--r-- 1 ant6 users  561 12-06 15:05 SOURCES.txt
-rw-r--r-- 1 ant6 users   13 12-06 15:05 top_level.txt

./tests:
total 32K
-rw-r--r-- 1 ant6 users    0 06-22 14:47 __init__.py
-rw-r--r-- 1 ant6 users  163 12-06 15:05 __init__.pyc
-rw-r--r-- 1 ant6 users 6,4K 12-06 14:49 test_curve.py
-rw-r--r-- 1 ant6 users 9,4K 12-06 15:05 test_curve.pyc
-rw-r--r-- 1 ant6 users  224 12-06 14:49 test_profile.py
-rw-r--r-- 1 ant6 users 1,2K 12-06 15:05 test_profile.pyc
```

**Comment by grzanka** [comment]
What if you use find_packages ?

**Comment by antnieszka** [comment]
For `packages=setuptools.find_packages(),`- the same, all 3 directories. I tired `bdist_wheel` and it and as zip with `beprof/, beprof-0.post0.dev........, tests/`. No *.pyc file were included in wheel.

**Comment by grzanka** [comment]
I see the point, it's not good to put tests at master level of site-packages (new issue can be created for pymchelper on that).
There are some reccomendations to put the tests inside the module directory: https://www.jeffknupp.com/blog/2013/08/16/open-sourcing-a-python-project-the-right-way/

Before proceeding try to do some reaseach on packaging formats and "where to put tests" issues.

We might also try to put in pypa:
 - wheel without tests
 - sdist which include tests

Check what happens if user is installing sdist using pip.

**Comment by antnieszka** [comment]
Ok, as for 'where to put tests' I found 2 common ways:
- putting tests into an extra directory outside your actual application code, useful if you have many functional tests or for other reasons want to keep tests separate from actual application code
- inlining test directories into your application package, useful if you have direct relation between (unit-)test and application modules and want to distribute your tests along with your application

Numpy as I see uses both approaches - some tests are inside packages core/ fft/ etc. (2nd point) and some are in external directory (1st point).

**Comment by grzanka** [comment]
OK, and what about what is included in various type of packages formats ?
Check for example this one: https://pypi.python.org/pypi/numpy -> there are two packages types: "Python Wheels" and "Source"

**Comment by antnieszka** [comment]
**numpy\*.whl**
Contains 3 dirs:
- beprof*-dist-info/
- beprof*-data/ - some additional scripts
- beprof/ - all numpy modules, tests, docs and more... looks like .zip of github repo
```
{16-12-06 17:53} ~/Downloads/numpy ant6% ➜ ls
add_newdocs.py  _distributor_init.py  fft               linalg      random
compat          distutils             _globals.py       ma          setup.py
__config__.py   doc                   _import_tools.py  matlib.py   testing
core            dual.py               __init__.py       matrixlib   tests
ctypeslib.py    f2py                  lib               polynomial  version.py
```
**source**
Similar:
```
(beprof){16-12-06 18:00} ~/Downloads/numpy-1.12.0b1 ant6% ➜ ll
total 56K
drwxr-xr-x  7 ant6 users  128 12-06 17:54 doc
-rw-rw-r--  1 ant6 users 5,2K 11-13 09:39 INSTALL.rst.txt
-rw-rw-r--  1 ant6 users 1,6K 11-13 09:39 LICENSE.txt
-rw-rw-r--  1 ant6 users 1,3K 11-13 09:39 MANIFEST.in
drwxr-xr-x 17 ant6 users 4,0K 12-06 17:54 numpy
-rw-rw-r--  1 ant6 users 2,3K 11-16 16:41 PKG-INFO
-rwxrwxr-x  1 ant6 users  15K 11-16 16:30 setup.py
-rw-rw-r--  1 ant6 users 8,5K 11-13 09:39 site.cfg.example
-rw-rw-r--  1 ant6 users 3,2K 07-09 22:12 THANKS.txt
drwxr-xr-x  3 ant6 users   18 12-06 17:54 tools
```
numpy/ also contains all modules and tests.

**Comment by antnieszka** [comment]
Also I created DataMedSci/pymchelper#148

**Comment by antnieszka** [comment]
After moving tests directory to beprof/:
```sh
(beprof3){16-12-06 18:12} ~/PycharmProjects/beprof/dist@feature/19-unit-testing✗✗✗✗✗✗ ant6% ➜ ll
total 16K
drwxr-xr-x 3 ant6 users 111 12-06 18:11 beprof
drwxr-xr-x 2 ant6 users 114 12-06 18:11 beprof-0.post0.dev1931481044288.dist-info
-rw-r--r-- 1 ant6 users 16K 12-06 18:11 beprof-0.post0.dev1931481044288-py2.py3-none-any.whl
(beprof3){16-12-06 18:12} ~/PycharmProjects/beprof/dist@feature/19-unit-testing✗✗✗✗✗✗ ant6% ➜ ls beprof
curve.py  functions.py  __init__.py  profile.py  tests  _version.py
```
code used in setup.py: `packages=setuptools.find_packages('.'),`

**Comment by grzanka** [comment]
So what is the difference between "Python wheel" and "Source" ?

**Comment by antnieszka** [comment]
"Python wheel" only contains code with tests and some metadata from wheel bdist build.
"Source" contains code with tests, documentation, some "set of files is for developing and testing file numpy" and files like LICENSE, README, INSTALL...

**Comment by grzanka** [comment]
Which one contains compiled C code and where ?

**Comment by antnieszka** [comment]
```py
    packages=setuptools.find_packages('.'),
    test_suite='beprof.tests',
```
`➜ python setup.py test` ran tests placed in beprof/tests.

As for C code: source code with header files and a lot more - are in "Source" version. Wheel contains only small portion of headers and mostly .py files.

**Comment by grzanka** [comment]
The whole idea behind serving users wheel files is to give them a package with python and compiled (!) C code. Most of the functionality of numpy package is written in C, python serves as a glue or envelope.
Think a bit of the concept of dtype, ufunc and memory allocation - how memory is allocated for normal python lists and how for numpy arrays (what happens if we append an item to both of them). Think about performance.

Can you locate compiled C code inside numpy wheel package (i.e. using `pip show -f numpy`) ?

It is also essential to see difference between universal and platform-dependent wheel. To which category numpy wheel belongs ?

**Comment by antnieszka** [comment]
Resolved in #110
